### PR TITLE
Use HashSet instead of Set in the type suggestion algorithm

### PR DESCRIPTION
**Pull request by vasily-kirichenko**
This (partially) fixes https://github.com/Microsoft/visualfsharp/issues/3882

### CODE REVIEW

**Comment by forki** [comment]
Test error looks like hash set has no stable order. But that's fine. I think changing the test assertion is OK since both values have same edit distance.

**Comment by vasily-kirichenko** [comment]
@dotnet-bot test Ubuntu14.04 Release_fcs Build please

**Comment by vasily-kirichenko** [comment]
It’s ready.

**Comment by vasily-kirichenko** [comment]
I also slightly optimized `jaro` function:

| Method |     Mean |     Error |    StdDev | Scaled |  Gen 0 | Allocated |
|------- |---------:|----------:|----------:|-------:|-------:|----------:|
|   Base | 502.9 ns | 9.5994 ns | 8.9793 ns |   1.00 | 0.2661 |    1120 B |
|    Opt | 341.0 ns | 0.7617 ns | 0.7125 ns |   0.68 | 0.1025 |     432 B |

**Comment by vasily-kirichenko** [comment]
| Method |     Mean |     Error |    StdDev | Scaled |  Gen 0 | Allocated |
|------- |---------:|----------:|----------:|-------:|-------:|----------:|
|   Base | 500.9 ns | 6.2139 ns | 5.5085 ns |   1.00 | 0.2661 |    1120 B |
|    Opt | 296.8 ns | 0.5939 ns | 0.5555 ns |   0.59 | 0.0644 |     272 B |

**Comment by vasily-kirichenko** [comment]
A bit more...

| Method |     Mean |     Error |    StdDev | Scaled |  Gen 0 | Allocated |
|------- |---------:|----------:|----------:|-------:|-------:|----------:|
|   Base | 494.3 ns | 0.7236 ns | 0.6768 ns |   1.00 | 0.2661 |    1120 B |
|    Opt | 246.3 ns | 1.8939 ns | 1.5815 ns |   0.50 | 0.0415 |     176 B |

**Comment by vasily-kirichenko** [comment]
| Method |     Mean |     Error |    StdDev | Scaled |  Gen 0 | Allocated |
|------- |---------:|----------:|----------:|-------:|-------:|----------:|
|   Base | 499.7 ns | 1.3196 ns | 1.1698 ns |   1.00 | 0.2661 |    1120 B |
|    Opt | 192.2 ns | 0.1997 ns | 0.1668 ns |   0.38 | 0.0417 |     176 B |

File: src/fsharp/NameResolution.fs [hunk 1]
```diff
@@ -2014,16 +2016,17 @@ let DecodeFSharpEvent (pinfos:PropInfo list) ad g (ncenv:NameResolver) m =
...
-        let typeName = NicePrint.minimalStringOfType nenv.eDisplayEnv typ
-        nenv.eFieldLabels
-        |> Seq.filter (fun kv -> 
-            kv.Value 
-            |> List.exists (fun r -> r.TyconRef.DisplayName = typeName))
-        |> Seq.map (fun kv -> kv.Key)
-        |> Set.ofSeq
-    else
-        Set.empty
+    (if isRecdTy g typ then
```
Comment (forki) [inline src/fsharp/NameResolution.fs]: Instead of () + pipe. Maybe introduce a temp variable

Comment (vasily-kirichenko) [reply src/fsharp/NameResolution.fs]: Yeah, I already didn't like it as I was typing it :) I just rewrote it with a CE.

File: src/fsharp/NameResolution.fs [hunk 2]
```diff
@@ -2014,16 +2016,12 @@ let DecodeFSharpEvent (pinfos:PropInfo list) ad g (ncenv:NameResolver) m =
...
-        let typeName = NicePrint.minimalStringOfType nenv.eDisplayEnv typ
-        nenv.eFieldLabels
-        |> Seq.filter (fun kv -> 
-            kv.Value 
-            |> List.exists (fun r -> r.TyconRef.DisplayName = typeName))
-        |> Seq.map (fun kv -> kv.Key)
-        |> Set.ofSeq
-    else
-        Set.empty
+    seq { if isRecdTy g typ then
```
Comment (forki) [inline src/fsharp/NameResolution.fs]: How about removing the seq {} and just use a for loop and imperative hashset.add?

Comment (vasily-kirichenko) [reply src/fsharp/NameResolution.fs]: Done.

**Pull request merged**